crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs

```diff
@@ -164,6 +175,87 @@ impl SpaceInternalHandler for Space {
         Ok(async move { Ok(res) }.boxed().into())
     }
 
+    fn handle_incoming_delegate_broadcast(
+        &mut self,
+        space: Arc<KitsuneSpace>,
+        basis: Arc<KitsuneBasis>,
+        _to_agent: Arc<KitsuneAgent>,
+        mod_idx: u32,
+        mod_cnt: u32,
+        data: crate::wire::WireData,
+    ) -> InternalHandlerResult<()> {
+        // first, forward this incoming broadcast to all connected
+        // local agents.
+        let mut local_events = Vec::new();
+        for agent in self.local_joined_agents.iter().cloned() {
+            let fut = self.evt_sender.notify(
+                space.clone(),
+                agent.clone(),
+                agent.clone(),
+                data.clone().into(),
+            );
+            local_events.push(async move {
+                if let Err(err) = fut.await {
+                    tracing::warn!(?err, "failed local broadcast");
+                }
+            });
+        }
+
+        // next, gather a list of agents covering this data to be
+        // published to.
+        let ro_inner = self.ro_inner.clone();
+        let timeout = ro_inner.config.tuning_params.implicit_timeout();
+        let fut =
+            discover::get_cached_remotes_near_basis(ro_inner.clone(), basis.get_loc(), timeout);
+
+        Ok(async move {
+            futures::future::join_all(local_events).await;
+
+            let info_list = fut.await?;
+
+            // for all agents in the gathered list, check the modulo params
+            // i.e. if `agent.get_loc() % mod_cnt == mod_idx` we know we are
+            // responsible for delegating the broadcast to that agent.
+            let mut all = Vec::new();
+            for info in info_list
+                .into_iter()
+                .filter(|info| info.agent.get_loc() % mod_cnt == mod_idx)
```

Comment:
I'm totally cool with this now that there are comments, but just an idea for the future that could enhance readability even more (not asking for this):

`mod_cnt` and `mod_idx` could be encapsulated in a struct like `DelegateParams {mod_cnt, mod_idx}`, and then Agent could have a method like `should_delegate(delegate_params: DelegateParams) -> bool`. Then the essence of conditional delegation is encapulated and there's more of a single focal point (the struct) for that aspect of publishing.

Reply:
I get where you're going from a code readability perspective... but from a serialization perspective, there's a tension in that the serialized api is effectively public.

The codec generation macro puts all the fields in a tuple so we don't waste space sending the field names over and over, and it's relatively simple at the moment, and doesn't support flattening.